Mathlib/Tactic/FastInstance.lean

```diff
@@ -0,0 +1,131 @@
+      Provided instance{indentExpr provided}\n\
+      is a proof, which does not need normalization."
+
+  -- Try to synthesize a total replacement for this term:
+  if let .some new ← trySynthInstance ty then
+    if ← withReducibleAndInstances <| isDefEq provided new then
+      trace[Elab.fast_instance] "replaced with synthesized instance"
+      return new
+    else
+      if ← withDefault <| isDefEq provided new then
+        error trace m!"\
+          Provided instance{indentExpr provided}\n\
+          is defeq only at default transparency to inferred instance{indentExpr new}"
+      else
+        error trace m!"\
+          Provided instance{indentExpr provided}\n\
+          is not defeq to inferred instance{indentExpr new}"
+  -- Otherwise, try to reduce it to a constructor.
+  else
+    -- Telescope since it might be a family of instances.
+    forallTelescopeReducing ty fun tyArgs _ => do
+      let provided' ← withReducibleAndInstances <| whnf <| mkAppN provided tyArgs
+      let error' (m : MessageData) : MetaM Expr := do
+        if isStructure (← getEnv) className then
+          error trace m
+        else
+          error trace m!"{m}\n\n\
+            This instance is not a structure and not canonical. \
+            Use a separate 'instance' command to define it."
+      let .const c .. := provided'.getAppFn
+        | error' m!"\
+            Provided instance does not reduce to a constructor application{indentExpr provided}"
+      let some (.ctorInfo ci) := (← getEnv).find? c
+        | error' m!"\
+            Provided instance does not reduce to a constructor application{indentExpr provided}\n\
+            Reduces to an application of {c}."
+      let mut args := provided'.getAppArgs
+      let params ← withDefault <| forallTelescopeReducing ci.type fun args _ =>
+        args.mapM fun arg => do
+          let recurse ← (return (← arg.fvarId!.getBinderInfo).isInstImplicit)
+                        <&&> not <$> Meta.isProof arg
+          return (recurse, ← arg.fvarId!.getUserName)
+      unless args.size == params.size do
+        -- This is an invalid term.
+        throwError "Incorrect number of arguments for constructor application{indentExpr provided'}"
+      for i in [ci.numParams:args.size] do
+        let (recurse, binderName) := params[i]!
+        if recurse then
+          let trace' := trace.push (className ++ binderName)
+          args := args.set! i (← makeFastInstance args[i]! (trace := trace'))
+      let provided' := mkAppN provided'.getAppFn args
+      mkLambdaFVars tyArgs provided'
+
+/--
+`fast_instance% inst` takes an expression for a typeclass instance `inst`, and unfolds it into
+constructor applications that leverage existing instances.
+
+For instance, when used as
+```lean
+instance instSemiring : Semiring X := sorry
+instance instRing : Ring X := fast_instance% Function.Injective.ring ..
+```
+this will define `instRing` as a nested constructor application that refers to `instSemiring`.
```

Comment:
This example seems useful, but please add a phrase about how this contrast with the usual situation. In other words, please finish the sentence with "as opposed to ..."

Reply:
Isn't this file already merged in master?

Reply:
Yes, but we can add this here if someone has a suggestion.

Reply:
I think it would be helpful to merge master to see what the current state of this file is in this PR.

Reply:
I am merging master and building locally at the moment. I believe it is exactly the same.

Reply:
Added.